pkg/blobinfocache/boltdb/boltdb.go

```diff
@@ -328,57 +345,72 @@ func (bdc *cache) CandidateLocations2(transport types.ImageTransport, scope type
 	var uncompressedDigestValue digest.Digest // = ""
+
+	getCandidatesForUncompressedDigest := func(tx *bolt.Tx, res []prioritize.CandidateWithTime, scopeBucket, compressionBucket *bolt.Bucket, primaryDigest digest.Digest, v2Output bool) ([]prioritize.CandidateWithTime, error) {
+		if uncompressedDigestValue = bdc.uncompressedDigest(tx, primaryDigest); uncompressedDigestValue != "" {
+			b := tx.Bucket(digestByUncompressedBucket)
+			if b != nil {
+				b = b.Bucket([]byte(uncompressedDigestValue.String()))
+				if b != nil {
+					if err := b.ForEach(func(k, _ []byte) error {
+						d, err := digest.Parse(string(k))
+						if err != nil {
+							return err
+						}
+						if d != primaryDigest && d != uncompressedDigestValue {
+							res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, d, v2Output)
+						}
+						return nil
+					}); err != nil {
+						return nil, err
+					}
+				}
+			}
+			if uncompressedDigestValue != primaryDigest {
+				res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, uncompressedDigestValue, v2Output)
+			}
+		}
+		return res, nil
+	}
+
 	if err := bdc.view(func(tx *bolt.Tx) error {
 		scopeBucket := tx.Bucket(knownLocationsBucket)
-		if scopeBucket == nil {
-			return nil
+		if scopeBucket != nil {
+			scopeBucket = scopeBucket.Bucket([]byte(transport.Name()))
 		}
-		scopeBucket = scopeBucket.Bucket([]byte(transport.Name()))
-		if scopeBucket == nil {
-			return nil
-		}
-		scopeBucket = scopeBucket.Bucket([]byte(scope.Opaque))
-		if scopeBucket == nil {
-			return nil
+		if scopeBucket != nil {
+			scopeBucket = scopeBucket.Bucket([]byte(scope.Opaque))
 		}
 		// compressionBucket won't have been created if previous writers never recorded info about compression,
 		// and we don't want to fail just because of that
 		compressionBucket := tx.Bucket(digestCompressorBucket)
 
-		res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, primaryDigest, requireCompressionInfo)
-		if canSubstitute {
-			if uncompressedDigestValue = bdc.uncompressedDigest(tx, primaryDigest); uncompressedDigestValue != "" {
-				b := tx.Bucket(digestByUncompressedBucket)
-				if b != nil {
-					b = b.Bucket([]byte(uncompressedDigestValue.String()))
-					if b != nil {
-						if err := b.ForEach(func(k, _ []byte) error {
-							d, err := digest.Parse(string(k))
-							if err != nil {
-								return err
-							}
-							if d != primaryDigest && d != uncompressedDigestValue {
-								res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, d, requireCompressionInfo)
-							}
-							return nil
-						}); err != nil {
-							return err
-						}
-					}
-				}
-				if uncompressedDigestValue != primaryDigest {
-					res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, uncompressedDigestValue, requireCompressionInfo)
+		// Add alls potential blobs with location unknown if v2Output is configured true.
```

Comment:
_Even further_:

Why is this calling `getCandidatesForUncompressedDigest` twice? If we have the `scopeBucket`, we don’t need unknown-location entries.

Is the only difference that we don’t call a`appendReplacementCandidates(… primaryDigest)` with `scopeBucket == nil`? If so, that isn’t a reason to duplicate the other code.

Also the `canSubstitute` logic seems incorrectly preserved: it is now not applied in the `scopeBucket == nil` variant.

AFAICS all of this code can be left as it is, maybe making the `primaryDigest` addition conditional. No need to factor out `getCandidatesForUncompressedDigest` function.

Reply:
@mtrmac We can reduce the call to one, but should we still factor out `getCandidatesForUncompressedDigest` just looks better to me for readability of the actual logic.

Reply:
(I’m not sure we actually envisage the same code operation yet)

The problem with factoring that out is:
- Primarily, naming and concept. E.g. it is not “for uncompressed digest”, the uncompressed digest is not a parameter. What does that function _do_, in isolation? If we can’t name it, it is IMHO a bad candidate for a function.
- Secondarily, the interface is a bit wide to be a clear-cut function. We can hide that in a nested closure, but it is still there.

The way I think about this, the `memory/cache.candidateLocations` is small and clear enough, and I think it does not need splitting. The thing that makes the BoltDB variant large is not the the logic is large, it’s mostly the mechanics of dealing with `digestByUncompressedBucket`. Maybe _that_ should be split into `cache.forEachDigestByUncompressedDigest(func(digest) error) error`; that would save two levels of nesting.

That seems worth doing, but it’s also a different PR>

Reply:
Fair enough I think actually the entire code here can stay as it is just ensuring that case `scopeBucket = nil` is also processed, however I'm just curious if this is possible for case of `memory` cause in memory there is nothing which actually verifies the passed in `scope` and sets it to `nil` so we will have to process for `nil` separately in case of memory.

Reply:
No: the equivalent of BoltDB’s `scopeBucket[scope] == nil` is Memory’s `mem.knownLocations[locationKey{…}] not found`. So setting `scope` to `nil` is not necessary.

Reply:
Ackd and pushed changes.